qiskit/circuit/quantumcircuit.py

```diff
@@ -3212,10 +3212,17 @@ def decompose(
         from qiskit.converters.dag_to_circuit import dag_to_circuit
 
         dag = circuit_to_dag(self, copy_operations=True)
-        dag = HighLevelSynthesis().run(dag)
+
+        if gates_to_decompose is None:
```

Comment:
Personally I would expect that an `MCXGate` or `Clifford` will be decomposed when I call `.decompose`, no?

Reply:
This is a bit of a corner-case. When a circuit has an `MCXGate` or a `Clifford`, the `decompose` function does decompose the object into lower-level gates. 

The question is what we should do when `decompose` is called with a non-`None` list of gates to decompose,  in which case it is expected that only the gates in this list will get decomposed but not any other gates (and in fact we have a test to check that an MCX gate will not get decomposed in this case). However, HighLevelSynthesis now does decompose MCXGates. So the change is that in this case we do not run `HighLevelSynthesis` prior to running `Decompose`. This is not a perfect fix, but at least avoids the failing testcases.

A better fix, as `HighLevelSynthesis` mostly subsumes the `Decompose` pass anyway is to add more options to `HighLevelSynthesis` dictating which gates to decompose and specifying the depth of the recursion.

Reply:
I see! It seems indeed that there's some duplicate functionality... let's discuss this in more detail, since HLS already does a lot of things I'm not sure we should add more 🤔